mathics/builtin/comparison.py

```diff
@@ -219,75 +219,66 @@ def numerify_args(items, evaluation):
-        ):
-            return lhs == rhs
-
-        # Dealing with two non-atomic expressions:
-        if not rhs.is_atom():
-            head1 = lhs.get_head()
-            head2 = rhs.get_head()
-
-            # FIXME: add equal2 method for DirectedInfinity
-
-            # Handling comparisons with DirectedInfinity
-            if head2.sameQ(SymbolDirectedInfinity):
-                lhs, rhs = rhs, lhs
-                head1, head2 = head2, head1
-            if head1.sameQ(SymbolDirectedInfinity):
-                if head2.sameQ(SymbolDirectedInfinity):
-                    dir1 = dir2 = Integer(1)
-                    if len(lhs._leaves) == 0:
-                        if len(rhs._leaves) == 0:
-                            return True
-                        dir1 = Integer(1)
-                    else:
-                        dir1 = lhs._leaves[0]
-                    if len(rhs._leaves) == 0:
-                        if dir1.sameQ(Integer(1)):
-                            return True
-                        dir2 = Integer(1)
-                    else:
-                        dir2 = rhs._leaves[0]
-                    # If the directions are equal,
-                    # then both infinites are the same
-                    if self.equal2(dir1, dir2):
-                        return True
-                    # Now, compare the signs:
-                    dir1 = Expression("Sign", dir1)
-                    dir2 = Expression("Sign", dir2)
-                    return self.equal2(dir1, dir2)
-
-        # Dealing with one expression
-        elif not lhs.is_atom():
-            # FIXME: Ass mentioned above: add equal2 method for DirectedInfinity
-            if lhs.get_head().sameQ(SymbolDirectedInfinity):
-                if isinstance(rhs, Number):
-                    return False
-                elif SymbolInfinity.sameQ(rhs):
-                    if len(lhs._leaves) == 0 or self.equal2(lhs._leaves[0], Integer(1)):
-                        return True
-
-        lhs_sympy = lhs.to_sympy(evaluate=True, prec=COMPARE_PREC)
-        rhs_sympy = rhs.to_sympy(evaluate=True, prec=COMPARE_PREC)
+    @staticmethod
+    def get_pairs(args):
+        for i in range(len(args)):
+            for j in range(i):
+                yield (args[i], args[j])
+
+    def expr_cmp(self, lhs, rhs, max_extra_prec=None):
+        if isinstance(rhs, Expression):
+            lhs, rhs = rhs, lhs
+        if not isinstance(lhs, Expression):
+            return
+        same_heads = lhs.get_head().sameQ(rhs.get_head())
+        if not same_heads:
+            return None
+        if len(lhs._leaves) != len(rhs._leaves):
+            return
+        for l, r in zip(lhs._leaves, rhs._leaves):
+            tst = self.equal2(l, r, max_extra_prec)
+            if not tst:
+                return tst
+        return True
+
+    def infty_cmp(self, lhs, rhs, max_extra_prec=None):
+        IntegerOne = Integer(1)
+        if rhs.get_head().sameQ(SymbolDirectedInfinity):
```

Comment:
Down the line I have in mind to reduce this kind of thing. But that can come later. 

I wonder though if some of this couldn't be simplified by defining an equal2 in DirectedInfinity.

Reply:
I tried to put this code in different places: in a ``DirectedInfinity`` subclass of ``Expression``, or of ``Atom`` or ``Number``, and as a method in the ``Builtin``, and as a rule in the corresponding builtin,  but at the end, this seems to be the proper place. 
In the first set of options, it would require to know ``max_extra_prec`` in the ``equal2`` method. As this information is stored in the ``Definitions`` object,  ``equal2`` should receive this information as an extra parameter. But for the other equal2 methods in ``BaseExpression`` superclasses, this parameter is not needed, so to add to ``Number`` and ``DirectedInfinity`` would introduce an odd inhomogeneity that I would prefer to avoid.

The other possibility is to add this as a method in the class ``DirectedInfinity`` superclass of  ``Builtin``. This could make sense if we consider to add this kind of methods on several other subclasses of ``Builtin``. Then, we should introduce a mechanism to look at all the Builtins to see if they have this method. But actually, we have this only example of a place where we need to implement this.
Finally, the other possibility would be to introduce Builtin rules like  ``DirectedInfinity[a_]==DirectedInfinity[b_] -> Sign[a]==Sign[b]`` and leave the parser to do the job.  This is not a bad option and is a valid way to make this extensible, relying on the existing infrastructure. However, I thought that it would be computationally faster to have these  Python methods on the ``_EqualityOperator``, and provides a concise and clear, and extensible way to deal with this class of special cases.

Reply:
> I tried to put this code in different places: in a `DirectedInfinity` subclass of `Expression`, or of `Atom` or `Number`, and as a method in the `Builtin`, and as a rule in the corresponding builtin, but at the end, this seems to be the proper place.
> In the first set of options, it would require to know `max_extra_prec` in the `equal2` method. As this information is stored in the `Definitions` object, `equal2` should receive this information as an extra parameter. But for the other equal2 methods in `BaseExpression` superclasses, this parameter is not needed, so to add to `Number` and `DirectedInfinity` would introduce an odd inhomogeneity that I would prefer to avoid.
> 
> The other possibility is to add this as a method in the class `DirectedInfinity` superclass of `Builtin`. This could make sense if we consider to add this kind of methods on several other subclasses of `Builtin`. Then, we should introduce a mechanism to look at all the Builtins to see if they have this method. But actually, we have this only example of a place where we need to implement this.
> Finally, the other possibility would be to introduce Builtin rules like `DirectedInfinity[a_]==DirectedInfinity[b_] -> Sign[a]==Sign[b]` and leave the parser to do the job. This is not a bad option and is a valid way to make this extensible, relying on the existing infrastructure. However, I thought that it would be computationally faster to have these Python methods on the `_EqualityOperator`, and provides a concise and clear, and extensible way to deal with this class of special cases.

Ok. Thanks for the information and trying. I guess we should just leave it and come back to it later.

As for the differences between equal2 routines, the conventional way to handle this is to add `*args, **kwargs` at the end of all of these functions. The ones that take a prec parameter would have `prec, *args, **kwargs`.  But if this is too much work. Leave it and we can come back later.